### PR TITLE
MXFP8 Online Quantization for Linear

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
### 1. vllm/model_executor/layers/quantization/mxfp8.py (NEW)
**Purpose:** Main MXFP8 quantization implementation for linear layers with torch.compile support.
**Key Components:**
| Component | Description |
|-----------|-------------|
| MXFP8_BLOCK_SIZE = 32 | Block size for E8M0 scales |
| E8M0_EXPONENT_BIAS = 127 | E8M0 format exponent bias |
| F8E4M3_MAX_POW2 = 8 | Largest power of 2 in FP8 E4M3 |
| mxfp8_quantize() | torch.compile-compatible quantization using bit manipulation |
| _to_blocked() | Rearranges scale matrix to cuBLAS blocked layout |
| Mxfp8Config | Quantization config class (requires SM100+) |
| Mxfp8LinearMethod | Linear layer quantization method |

**Key Methods:**
- `mxfp8_quantize()` - Quantizes tensor to FP8 E4M3 with E8M0 block scales using bit manipulation (no external dependencies, torch.compile compatible)
- `_to_blocked()` - Converts scale matrix to cuBLAS blocked layout per NVIDIA spec, with `torch.compiler.is_compiling()` check for compile compatibility
- `create_weights()` - Creates weight parameters in bfloat16 for online quantization
- `process_weights_after_loading()` - Quantizes weights to FP8 E4M3 with pre-blocked E8M0 scales
- `apply()` - Dynamically quantizes activations, pads M dimension for bias support, performs matmul via torch._scaled_mm

**Dependencies:**
- torch._scaled_mm with block scales (E8M0 format)
- No external quantization library dependencies (replaced flashinfer with custom implementation)

---

### 2. vllm/model_executor/layers/quantization/__init__.py (MODIFIED)
**Changes:**
- Added "mxfp8" to QuantizationMethods literal type
- Added import: `from .mxfp8 import Mxfp8Config`
- Added mapping: `"mxfp8": Mxfp8Config`

---

### 3. tests/quantization/test_mxfp8.py (NEW)

**Tests:**

| Test | Description |
|------|-------------|
| test_load_bf16_model | End-to-end test loading facebook/opt-125m with MXFP8 quantization |
| test_mxfp8_linear_method_apply | Unit test for Mxfp8LinearMethod.apply() numerical correctness |

**Test Parameters:**
- `test_load_bf16_model` parametrized with `enforce_eager=[True, False]` (tests both eager and compile modes)
- `test_mxfp8_linear_method_apply` parametrized with `use_bias=[False, True]`
- Uses SQNR > 15 dB threshold for numerical correctness validation
- Skip Condition: Tests skip if GPU capability < SM100 (Blackwell)

**Helper Functions:**
- `compute_sqnr()` - Computes Signal-to-Quantization-Noise Ratio in dB
---

### Technical Details

<details>

```
Quantization Flow:
BF16 weights → mxfp8_quantize() → FP8 E4M3 + E8M0 scales
                                          ↓
                    scales → _to_blocked() → cuBLAS blocked layout (pre-computed)
                                          ↓
                    stored as layer.weight (FP8) + layer.weight_scale (blocked E8M0)
BF16 activations → pad M to 32 → mxfp8_quantize() → FP8 E4M3 + E8M0 scales
                                          ↓
                    scales → _to_blocked() → cuBLAS blocked layout (runtime)
                                          ↓
                    torch._scaled_mm(x_fp8, weight_fp8.t(),
                                    x_scale_blocked, w_scale_blocked,
                                    bias=bias, out_dtype=bf16)
                                          ↓
                    output → remove padding → view to original shape
                    
```
                    
Matrix Layout Requirements:
- Input (a): [M_padded, K] row-major (contiguous), M_padded = max(M, 32)
- Weight (b): [N, K] row-major → .t() → [K, N] column-major
- Input scales: [M_padded, K//32] → _to_blocked() → flattened cuBLAS format (runtime)
- Weight scales: [N, K//32] → _to_blocked() → flattened cuBLAS format (pre-computed)
- Bias: Passed directly to _scaled_mm (requires M >= 17, hence padding to 32)

torch.compile Compatibility:
- mxfp8_quantize() uses pure PyTorch ops with bit manipulation
- _to_blocked() has torch.compiler.is_compiling() check for consistent tensor creation
- M padding uses torch.compiler.is_compiling() check for consistent control flow
- Weight scales pre-blocked during loading to avoid runtime blocking under compile

```

</details>


**Requirements:**
- GPU: SM100+ (Blackwell)
- PyTorch: 2.8.0+ (for torch._scaled_mm with E8M0 block scales)
- Input dtype: bfloat16 only
- Online quantization only (no pre-quantized checkpoints)
**Usage:**
```bash
# With FLASH_ATTN backend (recommended for torch.compile)
VLLM_ATTENTION_BACKEND=FLASH_ATTN VLLM_DISABLE_COMPILE_CACHE=1 vllm serve <model> --dtype bfloat16 --quantization mxfp8
# Eager mode (works with any attention backend)
vllm serve <model> --dtype bfloat16 --quantization mxfp8 --enforce-eager
```

---
Test Commands

pytest tests/quantization/test_mxfp8.py

Result

<details>

```
============================= test session starts ==============================
platform linux -- Python 3.12.0, pytest-9.0.2, pluggy-1.6.0
rootdir: /home/jerryzh/local/vllm
configfile: pyproject.toml
plugins: anyio-4.11.0
collected 4 items

tests/quantization/test_mxfp8.py ....                                    [100%]

=============================== warnings summary ===============================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

tests/quantization/test_mxfp8.py::test_load_bf16_model[True]
tests/quantization/test_mxfp8.py::test_load_bf16_model[False]
  /home/jerryzh/.conda/envs/vllm/lib/python3.12/multiprocessing/popen_fork.py:66: DeprecationWarning: This process (pid=4157396) is multi-threaded, use of fork() may lead to deadlocks in the child.
    self.pid = os.fork()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================== 4 passed, 4 warnings in 40.00s ========================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
```

```
VLLM_ATTENTION_BACKEND=FLASH_ATTN VLLM_DISABLE_COMPILE_CACHE=1 with-proxy vllm serve facebook/opt-125m --quantization mxfp8 --dtype bfloat16
```

<details>

Note that there are some pre-existing error the default flashinfer (without quantization), that's why we set attention backend to FLASH_ATTN:
```
(EngineCore_DP0 pid=3024207)   File "/home/jerryzh/local/vllm/vllm/v1/attention/backends/flashinfer.py", line 1507, in forward
(EngineCore_DP0 pid=3024207)     assert is_strictly_contiguous(decode_query)
(EngineCore_DP0 pid=3024207) AssertionError
[rank0]:[W114 17:28:57.561665561 ProcessGroupNCCL.cpp:1524] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs
/stable/distributed.html#shutdown (function operator())
```

```
...
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /v1/messages, Methods: POST
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /v1/completions, Methods: POST
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /ping, Methods: GET
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /ping, Methods: POST
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /invocations, Methods: POST
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /classify, Methods: POST
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /v1/embeddings, Methods: POST
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /score, Methods: POST
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /v1/score, Methods: POST
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /rerank, Methods: POST
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /v1/rerank, Methods: POST
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /v2/rerank, Methods: POST
(APIServer pid=3525388) INFO 01-14 17:41:04 [launcher.py:46] Route: /pooling, Methods: POST
(APIServer pid=3525388) INFO:     Started server process [3525388]
(APIServer pid=3525388) INFO:     Waiting for application startup.
(APIServer pid=3525388) INFO:     Application startup complete.
^C[rank0]:[W114 17:41:36.897933262 ProcessGroupNCCL.cpp:1524] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
(APIServer pid=3525388) INFO 01-14 17:41:36 [launcher.py:110] Shutting down FastAPI HTTP server.
(APIServer pid=3525388) INFO:     Shutting down
(APIServer pid=3525388) INFO:     Waiting for application shutdown.
(APIServer pid=3525388) INFO:     Application shutdown complete.
```

</details>

Performance

<details>

CUDA_VISIBLE_DEVICES=1 VLLM_ATTENTION_BACKEND=FLASH_ATTN VLLM_DISABLE_COMPILE_CACHE=1 with-proxy vllm bench latency --input-len 1024 --output-len 1024 --model Qwen/Qwen3-8B --batch-size 32 --quantization mxfp8 
Avg latency: 7.114737543479229 seconds
10% percentile latency: 7.090550634596729 seconds
25% percentile latency: 7.095485741025186 seconds
50% percentile latency: 7.100555320997955 seconds
75% percentile latency: 7.107174740507617 seconds
90% percentile latency: 7.116721842222614 seconds
99% percentile latency: 7.380051661875914 seconds

CUDA_VISIBLE_DEVICES=1 VLLM_ATTENTION_BACKEND=FLASH_ATTN VLLM_DISABLE_COMPILE_CACHE=1 with-proxy vllm bench latency --input-len 1024 --output-len 1024 --model Qwen/Qwen3-8B --batch-size 32

Avg latency: 6.2698817714350294 seconds
10% percentile latency: 6.262251399946399 seconds
25% percentile latency: 6.264382338747964 seconds
50% percentile latency: 6.268500665522879 seconds
75% percentile latency: 6.271425302984426 seconds
90% percentile latency: 6.277459568175255 seconds
99% percentile latency: 6.29489082104119 seconds

CUDA_VISIBLE_DEVICES=1 VLLM_ATTENTION_BACKEND=FLASH_ATTN VLLM_DISABLE_COMPILE_CACHE=1 with-proxy vllm bench latency --input-len 1024 --output-len 1024 --model Qwen/Qwen3-8B --batch-size 8

Avg latency: 5.377933108761984 seconds
10% percentile latency: 5.36727682906785 seconds
25% percentile latency: 5.369608670487651 seconds
50% percentile latency: 5.370721633022185 seconds
75% percentile latency: 5.374492640490644 seconds
90% percentile latency: 5.382614015077706 seconds
99% percentile latency: 5.4619856172625445 seconds

CUDA_VISIBLE_DEVICES=1 VLLM_ATTENTION_BACKEND=FLASH_ATTN VLLM_DISABLE_COMPILE_CACHE=1 with-proxy vllm bench latency --input-len 1024 --output-len 1024 --model Qwen/Qwen3-8B --batch-size 8 --quantization mxfp8

Avg latency: 5.8796797158953265 seconds
10% percentile latency: 5.868537089729216 seconds
25% percentile latency: 5.873511790239718 seconds
50% percentile latency: 5.876653404004173 seconds
75% percentile latency: 5.885830606741365 seconds
90% percentile latency: 5.8914141334069425 seconds
99% percentile latency: 5.9247187449870395 seconds
[rank0]:[W204 18:49:56.613028276 ProcessGroupNCCL.cpp:1524] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())

——————
CUDA_VISIBLE_DEVICES=1 VLLM_ATTENTION_BACKEND=FLASH_ATTN VLLM_DISABLE_COMPILE_CACHE=1 with-proxy vllm bench latency --input-len 1024 --output-len 1024 --model Qwen/Qwen3-32B --batch-size 256 --quantization mxfp8
Avg latency: 53.336440228034434 seconds
10% percentile latency: 53.07816166278208 seconds
25% percentile latency: 53.27027780821663 seconds
50% percentile latency: 53.39790177551913 seconds
75% percentile latency: 53.4346607557527 seconds
90% percentile latency: 53.45001104929834 seconds
99% percentile latency: 53.49330673501012 seconds

CUDA_VISIBLE_DEVICES=1 VLLM_ATTENTION_BACKEND=FLASH_ATTN VLLM_DISABLE_COMPILE_CACHE=1 with-proxy vllm bench latency --input-len 1024 --output-len 1024 --model Qwen/Qwen3-32B --batch-size 256
Avg latency: 75.82325858416735 seconds
10% percentile latency: 75.72335491161793 seconds
25% percentile latency: 75.74582710850518 seconds
50% percentile latency: 75.86072600752232 seconds
75% percentile latency: 75.88143472374941 seconds
90% percentile latency: 75.89680293400306 seconds
99% percentile latency: 75.91749269683962 seconds
[rank0]:[W204 20:40:52.323331938 ProcessGroupNCCL.cpp:1524] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
</details>

---
Usage Example

from vllm import LLM

llm = LLM(
    model="facebook/opt-125m",
    quantization="mxfp8",
    dtype="bfloat16",
)

Generated by claude

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/mxfp8.py [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+MBITS_F32 = 23  # mantissa bits in float32
+
+
+def _ceil_div(a: int, b: int) -> int:
+    """Ceiling division."""
+    return (a + b - 1) // b
+
+
+# from https://github.com/pytorch/ao/blob/c6d21f9b82819a7e3546a267eafb15a5b8cf3951/torchao/prototype/mx_formats/mx_tensor.py#L144
+def mxfp8_quantize(
```
Comment (vkuzo) [inline vllm/model_executor/layers/quantization/mxfp8.py]: i think is complex enough to go in a separate file

File: vllm/model_executor/layers/quantization/mxfp8.py [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+    data = data.reshape(*orig_shape[:-1], orig_shape[-1] // block_size, block_size)
+
+    # Find max absolute value per block
+    max_abs = torch.amax(torch.abs(data), dim=-1, keepdim=True)
+
+    # Cast to float32 for scale calculation (required for bit manipulation)
+    data = data.to(torch.float32)
+    max_abs = max_abs.to(torch.float32)
+
+    # Extract the exponent from max_abs using bit manipulation
```
Comment (vkuzo) [inline vllm/model_executor/layers/quantization/mxfp8.py]: this should be using the RCEIL scale calculation method for improved accuracy: https://github.com/pytorch/ao/blob/f222b9ed43f4254a12ad291ac6718c0f8d779cdb/torchao/prototype/mx_formats/mx_tensor.py#L96

File: vllm/model_executor/layers/quantization/mxfp8.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+
+        # Transform input scale to blocked layout for cuBLAS
+        x_scale_blocked = _to_blocked(x_scale_2d)
+
+        # Use torch._scaled_mm with block scales
+        # a (x_fp8): [M_padded, K] row-major (contiguous)
+        # b (weight_fp8.t()): [K, N] column-major (weight_fp8 is [N, K] row-major,
+        #   so weight_fp8.t() is [K, N] with stride [1, K] = column-major)
+        output = torch._scaled_mm(
+            x_fp8,
```
Comment (vkuzo) [inline vllm/model_executor/layers/quantization/mxfp8.py]: nit: extract x -> x_fp8, x_scale_blocked into a utility function instead of inlining

File: vllm/model_executor/layers/quantization/mxfp8.py [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+            weight_fp8.t(),  # [K, N] column-major
+            x_scale_blocked.view(torch.float8_e8m0fnu),
```
Comment (vkuzo) [inline vllm/model_executor/layers/quantization/mxfp8.py]: function that calculates scale should return `e8m0fnu` dtype

**Comment by jerryzh168** [comment]
1. not tested yet, will test now since the overall approach looks good to you
2. I heard we already have mxfp8 scaled_grouped_mm support in pytorch core, going to test that out as a first step (https://docs.pytorch.org/docs/main/generated/torch.nn.functional.scaled_grouped_mm.html), Simon is also working on moving a CuteDSL implementation of it to pytorch core, but that's probably an implementation detail

**Comment by jerryzh168** [comment]
no longer needed since https://github.com/vllm-project/vllm/pull/35448 is landed
